docs/source/development/eep-02-typing.md

```diff
@@ -0,0 +1,1474 @@
+   autocomplete is not very helpful. Instead, the user would have to be able to filter
+   the autocomplete results according to the problem properties (e.g. least-squares,
+   gradient-based, local, ...). However, it is not clear which filters are relevant and
+   in which order a user wants to apply them.
+
+#### Proposal
+
+We continue to support passing algorithms as strings. This is important because
+estimagic promises to work "just like SciPy" for simple things. On top, we offer a new
+way of specifying algorithms that is less prone to typos, supports autocomplete and will
+be useful for advanced algorithm configuration.
+
+To exemplify the new approach, assume a simplified situation with 5 algorithms. We only
+consider whether an algorithm is gradient free or gradient based. One algorithm is not
+installed, so should never show up anywhere. Here is the fictitious list:
+
+- `neldermead`: installed, `gradient_free`
+- `bobyqa`: installed, `gradient_free`
+- `lbfgs`: installed, `gradient_based`
+- `slsqp`: installed, `gradient_based`
+- `ipopt`: not installed, `gradient_based`
+
+We want the following behavior:
+
+The user types `em.algorithms.` and autocomplete shows
+
+|                 |
+| --------------- |
+| `GradientBased` |
+| `GradientFree`  |
+| `neldermead`    |
+| `bobyqa`        |
+| `lbfgs`         |
+| `slsqp`         |
+
+A user can either select one of the algorithms (lowercase) directly or filter further by
+selecting a category (CamelCase). This would look as follows:
+
+The user types `em.algorithms.GradientFree.` and autocomplete shows
+
+|              |
+| ------------ |
+| `neldermead` |
+| `bobyqa`     |
+
+Once the user arrives at an algorithm, a subclass of `Algorithm` is returned. This class
+will be passed to `minimize` or `maximize`. Passing configured instances of `Algorithm`s
+will be discussed in [Algorithm Options](algorithm-options).
+
+In practice, we would have a lot more algorithms and a lot more categories. Some
+categories might be mutually exclusive, in that case the second category is omitted
+after the first one is selected.
+
+We have the following categories:
+
+- `GradientBased` vs. `GradientFree`
+- `Local` vs. `Global`
+- `Bounded` vs. `Unbounded`
+- `Scalar` vs. `LeastSquares` vs. `Likelihood`
+- `LinearConstrained` vs. `NonlinearConstrained` vs. `Unconstrained`
+
+Potentially, we could also offer a `.All` attribute that returns a list of all currently
+selected algorithms. That way a user could for example loop over all `Bounded` and
+`GradientBased` `LeastSquares` algorithms and compare them in a criterion plot.
+
+These categories match nicely with our
+[algorithm selection tutorials](https://effective-programming-practices.vercel.app/scientific_computing/optimization_algorithms/objectives_materials.html).
+
+We can use
+[`dataclasses.make_dataclass`](https://docs.python.org/3/library/dataclasses.html#dataclasses.make_dataclass)
```

Comment:
Hey estimagic team, nice proposal. I had a discussion with @janosg about this part. Using `make_dataclass` will not give static type support, as the type is build dynamically at runtime.

This feature is a great idea, but my guess is it will take some effort. I would tackle it the following way:

1. For every category, add a class attribute to the `Algorithm` base class (this will just be a stub)
    ```python
     class Algorithm(ABC):
         is_gradient_based: bool
         ....
    ```
2. Use `__init__subclass__` to check that a subclass sets the attribute (also validate it), e.g.:
    ```python
     class Algorithm(ABC):
         is_gradient_based: bool
         def __init_subclass__(cls, **kwargs):
                if not hasattr(cls, "is_gradient_based"):
                    raise TypeError(...)
        
                if not isinstance(cls.is_gradient_based, bool):
                    raise ValueError(...)
                
                super().__init_subclass__(**kwargs)
    ```
3. Write a script, which search for subclasses of `Algorithm` and based on the class attributes builds a static data structure as code representing the wanted tree (I mean a script which generates the corresponding python code and writes it to a file). Instead of searching, one could also register the subclasses in `__init__subclas__`
4. Add a pre-commit hook, which calls the file generating script and validates the output.
5. Add tests to ensure the generated code is valid python code (e.g. correct indentation, etc.)

Reply:
Hi @schroedk, welcome to estimagic and thanks for pointing this out. This is completely correct, my approach only worked in jupyter notebooks, which is not good enough. 

I think your proposal contains two separate points:
1. An approach to use a pre-commit hook to achieve static analysis and autocomplete for algorithm selection as well as filtering algorithms by category. That approach will also work in all other places where I suggested to use `make_dataclass`.
2. An alternative to the `mark.minimizer` decorator via class variables that are validated via `__init_subclass__`. 

I am completely convinced about the first point and will adjust my proposal. 

I will also add the second point as alternative to the decorator approach so we can have a discussion about the pros and cons of each idea. My main concern is that most people who have contributed algorithms to estimagic so far do not know the difference between class and instance variables but are very familiar with using decorators. It will be interesting to hear opinions on this.